### PR TITLE
Rnr 115 sanctuary tutorial

### PR DESCRIPTION
**Pull request by jack829**
## JIRA Link
https://scrychat.atlassian.net/browse/RNR-115

## Changes
* Refactors https://github.com/TryScry/reactNativePOC/pull/72 to add a sibling Overlay element next to the Router. These are both contained in a component that passes down a `routeListener` method to be called in the `onEnter` method of each scene. Here, the scene can pass back the `routeName` and the `OverlayContainer` can determine which tutorial needs to be shown, if any at all
* add new HOC `addSetElementDimensions` that connects to the store and handles parsing the `onLayout` event and saving relevant ui element dimensions
* Add `OverlayContainer` that right now contains all of the data about each tutorial (this replaces `ScryChatListNav.container`)-> how many parts, the modal text for each part, and the sections to highlight for each part. Also move `ScryChatListNav` to `TutorialOverlay` (sorry, I fucked this one up git-wise so it looks brand new).
* Refactor `ChatInput` and `ScryingRoomButton` to their own components that can be re-rendered for the tutorial

## Other Dependencies
none

## ScreenShots
rnr115-sanctuary-tut-1
rnr115-sanctuary-tut-2
rnr115-santuary-tut-3

### CODE REVIEW

File: src/screens/ScryChat/ScryChat.js [hunk 1]
```diff
@@ -20,8 +20,15 @@ import AppText from '../../components/AppText'
 import ScryChatMessage from './ScryChatMessage'
 import TinderChatEmbedded from '../TinderChat/TinderChatEmbedded'
 import SuggestionStatus from '../../components/SuggestionStatus'
+import ChatInput from '../../components/ChatInput'
+import addSetElementDimensionsWrapper from '../../HOC/addSetElementDimensionsWrapper';
```
Comment (dblanchard13) [inline src/screens/ScryChat/ScryChat.js]: NIT: `;`

File: src/screens/ScryChatList/ScryChatList.container.js [hunk 2]
```diff
@@ -24,6 +24,7 @@ import {
   getUnseenFriendsScryChats,
 } from '../../store/scryChats/selectors'
 import ScryChatList from './ScryChatList'
+import { tutorialMap } from '../../globals/tutorials';
```
Comment (dblanchard13) [inline src/screens/ScryChatList/ScryChatList.container.js]: NIT: `;`

File: src/screens/ScryChatList/ScryChatListSubHeader.js [hunk 3]
```diff
@@ -18,11 +18,9 @@ class ScryChatListSubHeader extends Component  {
   friendsChatsToggleOffset = 0
 
   onLayout = (event, toggleName, otherProps) => {
-
     if (this.props.completedChatsPageTutorial || this[toggleName]) return
-
+    // Native event must be pulled out synchronously
```
Comment (dblanchard13) [inline src/screens/ScryChatList/ScryChatListSubHeader.js]: Could you update this to `nativeEvent` for clarity and consistency with the below comment?

File: src/screens/TinderChat/TinderChatEmbedded.js [hunk 4]
```diff
@@ -32,7 +32,7 @@ class TinderChatEmbedded extends Component {
   }
 
   render() {
-    const { match_thumbnail_pic: thumbNail, title } = this.props.conversation
+    const { match_thumbnail_pic: thumbNail , title } = this.props.conversation
```
Comment (dblanchard13) [inline src/screens/TinderChat/TinderChatEmbedded.js]: Would prefer you not have this space..

File: src/HOC/addOverlayWrapper.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    routeListener = (props) => {
+      const { routeName } = props
+      if (routeTutorialMap[routeName]) {
+        this.setState({ routeName })
+      }
+    }
+
+    render() {
+      return (
+        <React.Fragment>
```
[the remark this answers is not in the document]
Comment (jack829) [reply src/HOC/addOverlayWrapper.js]: Na, I just saw it like this once and didn't even think about it

File: src/components/Backdrop.js [hunk 6]
```diff
@@ -13,15 +13,21 @@ const {
 
 function Backdrop ({ visible = false, highlightSections }) {
   let positionsLoaded = true
-  highlightSections.forEach((section) => {
-    if (!section || !section.component) positionsLoaded = false
-  })
+  highlightSectionsArr = []
```
[the remark this answers is not in the document]
Comment (jack829) [reply src/components/Backdrop.js]: whoa, fucking god dammit. This is that shit where I'm just not getting errors

File: src/components/Backdrop.js [hunk 7]
```diff
@@ -59,7 +65,7 @@ function renderHighlight(section) {
     top: y - 35,
     left: x - 25,
   }
-
+  console.log()
```
[the remark this answers is not in the document]
Comment (jack829) [reply src/components/Backdrop.js]: oops

File: src/components/ScryingRoomButton.js [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+import {
+  View,
+  TouchableOpacity,
+  Image,
+  Dimensions,
+} from 'react-native'
+import { Actions } from 'react-native-router-flux'
+
+import {
+  NAV_CHEVRON, COLOR_PRIMARY,
```
[the remark this answers is not in the document]
Comment (jack829) [reply src/components/ScryingRoomButton.js]: ah, the auto import

File: src/screens/TinderChat/TinderChatEmbedded.js (hunk 4 above)
Comment (jack829) [reply src/screens/TinderChat/TinderChatEmbedded.js]: fuck you

**Pull request merged**